performance/gcd_bench.cpp

```diff
@@ -0,0 +1,215 @@
+                  u = eval_gcd(*v.limbs(), *u.limbs());
+               else
+               {
+                  double_limb_type i = v.limbs()[0] | (static_cast<double_limb_type>(v.limbs()[1]) << sizeof(limb_type) * CHAR_BIT);
+                  double_limb_type j = (u.size() == 1) ? *u.limbs() : u.limbs()[0] | (static_cast<double_limb_type>(u.limbs()[1]) << sizeof(limb_type) * CHAR_BIT);
+                  u                  = eval_gcd(i, j);
+               }
+               break;
+            }
+            //
+            // Regular binary gcd case:
+            //
+            eval_subtract(v, u);
+            vs = eval_lsb(v);
+            eval_right_shift(v, vs);
+         } while (true);
+
+         result = u;
+         eval_left_shift(result, shift);
+      }
+
+      }
+   }
+}
+
+template <class T>
+std::tuple<std::vector<T>, std::vector<T>, std::vector<T> >& get_test_vector(unsigned bits)
+{
+   static std::map<unsigned, std::tuple<std::vector<T>, std::vector<T>, std::vector<T> > > data;
+
+   std::tuple<std::vector<T>, std::vector<T>, std::vector<T> >& result = data[bits];
+
+   if (std::get<0>(result).size() == 0)
+   {
+      mt19937                     mt;
+      uniform_int_distribution<T> ui(T(1) << (bits - 1), T(1) << bits);
+
+      std::vector<T>& a = std::get<0>(result);
+      std::vector<T>& b = std::get<1>(result);
+      std::vector<T>& c = std::get<2>(result);
+
+      for (unsigned i = 0; i < 1000; ++i)
+      {
+         a.push_back(ui(mt));
+         b.push_back(ui(mt));
+         if (b.back() > a.back())
+            b.back().swap(a.back());
+         c.push_back(0);
+      }
+   }
+   return result;
+}
+
+template <class T>
+std::vector<T>& get_test_vector_a(unsigned bits)
+{
+   return std::get<0>(get_test_vector<T>(bits));
+}
+template <class T>
+std::vector<T>& get_test_vector_b(unsigned bits)
+{
+   return std::get<1>(get_test_vector<T>(bits));
+}
+template <class T>
+std::vector<T>& get_test_vector_c(unsigned bits)
+{
+   return std::get<2>(get_test_vector<T>(bits));
+}
+
+
+template <typename T>
+static void BM_gcd_old(benchmark::State& state)
+{
+   int                         bits = state.range(0);
+
+   std::vector<T>& a = get_test_vector_a<T>(bits);
+   std::vector<T>& b = get_test_vector_b<T>(bits);
+   std::vector<T>& c = get_test_vector_c<T>(bits);
+
+   for (auto _ : state)
+   {
+      for (unsigned i = 0; i < a.size(); ++i)
+         eval_gcd_old(c[i].backend(), a[i].backend(), b[i].backend());
+   }
+   state.SetComplexityN(bits);
+}
+
+template <typename T>
+static void BM_gcd_current(benchmark::State& state)
+{
+   int                         bits = state.range(0);
+
+   std::vector<T>& a = get_test_vector_a<T>(bits);
+   std::vector<T>& b = get_test_vector_b<T>(bits);
+   std::vector<T>& c = get_test_vector_c<T>(bits);
+
+   for (auto _ : state)
+   {
+      for (unsigned i = 0; i < a.size(); ++i)
+         eval_gcd(c[i].backend(), a[i].backend(), b[i].backend());
```

Comment:
Wait, why are you doing an eval_gcd on the entire vector? Why not just do it on two single numbers?

Reply:
GCD performance is highly dependent on the inputs, you want to try and get a decent average.  To pick a trivial example, if one of the numbers has a lot of consecutive zero bits, then binary GCD would look way better than it actually is.